web3/utils/datatypes.py

```diff
@@ -0,0 +1,50 @@
+from functools import (
+    partial,
+)
+from eth_utils import (
+    to_dict,
+)
+from toolz.itertoolz import (
+    concat
+)
+
+
+def verify_attr(class_name, key, base):
+    if not hasattr(base, key):
+        raise AttributeError(
+            "Property {0} not found on {1} class. "
+            "`{1}.factory` only accepts keyword arguments which are "
+            "present on the {1} class".format(key, class_name)
+        )
+
+
+class PropertyCheckingFactory(type):
+    @to_dict
+    def normalize_properties(attributes, values, normalizers):
+        for attribute, value in zip(attributes, values):
+            if attribute in normalizers:
+                normalizer = normalizers[attribute]
+                yield attribute, normalizer(value)
+            else:
+                yield attribute, value
+
+    def __init__(cls, name, bases, namespace, **kargs):
+        # see PEP487.  To accept kwargs in __new__, they need to be
+        # filtered out here.
+        super().__init__(name, bases, namespace)
+
+    def __new__(mcs, name, bases, namespace, normalizers={}):
+        for key in namespace:
+            verify_key_attr = partial(verify_attr, name, key)
+            map(verify_key_attr, set(concat(base.__mro__ for base in bases)))
+
+        attributes, values = zip(*namespace.items())
```

Comment:
I find it somewhat odd that `normalize_properties` takes the attrs and values as separate arguments.

First, *if* attributes and values were of different lengths then the extra ones from whichever was longer would be silently discarded.  This could be fixed by passing them in as a single iterable of 2-tuples.  `(('key-1', 'value-a'), ('key-2', 'value-2'), ...)`

Second, It seems even cleaner to just pass the `namespace` directly in and change it to be `mcs.normalize_properties(namespace, normalizers)` and then do the `items()` call within that function.

And last, `mcs.normalize_properties` could be replaced with `web3.utils.formatters.apply_formatters_to_dict`.
- it seems like we should be able to instea

Reply:
Castle of Aaaaargh?

Seems like the last comment here got cut off.

Reply:
i hope he's ok

Reply:
No clue